### PR TITLE
Create new imagery provider for MapBox styles

### LINKED ISSUE #4015: Mapbox Classic will be deprecated
Hi, I have a project using "classic" Mapbox imagery layers, which will be deprecated,
¿Is there any plans to migrate to the new Mapbox?

### PR DESCRIPTION
**Pull request by bampakoa**
This work intends to incorporate the new Mapbox styles API into Cesium.

Fixes #4015

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @bampakoa!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/mapbox-styles/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

**Comment by hpinkos** [comment]
Great, thanks @bampakoa!  We'll take a look at this after the release goes out

**Comment by bampakoa** [comment]
@hpinkos this is a draft PR yet. I plan to push some more code along with tests and README changes this week and then I can notify you for reviewing 😉, if that is ok for you.

**Comment by hpinkos** [comment]
Oh yes!  I missed that.  Yeah, let me know when its ready =)

**Comment by bampakoa** [comment]
@hpinkos this is ready for review now. You can check it directly in the Hello World app by selecting the Mapbox layers from the picker (I guess I should revert it back l8r to the old Mapbox 😃 ). I have not updated the CHANGES.md file yet since there is not any relevant entry for the next release 😟 

@cesium-concierge I have updated the tests 😉

**Comment by mrlubos** [comment]
Thanks for making this pull request @bampakoa ! ❤️

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @bampakoa! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by hpinkos** [comment]
@OmarShehata can you review?

**Comment by OmarShehata** [comment]
Thanks for putting this together @bampakoa ! It seems to work pretty well. I do have some notes/questions:

* Can you merge in master and push?
* Why is it called MapboxStyleProvider instead of MapboxStyleImageryProvider? It _is_ an imagery provider, so it should be consistent with the other imagery providers listed here https://cesiumjs.org/Cesium/Build/Documentation/ImageryProvider.html?classFilter=ImageryProvid
* It should probably be added to that list too I think ^
* Since the old Mapbox API is deprecated, is there a reason to keep around the old MapboxImageryProvider? Can this new class just replace the old one? 
  * I think this would be a breaking change in that the user needs to change the parameters they pass to the constructor, to choose which style they want, right?

**Comment by mrlubos** [comment]
@OmarShehata Hey, to the last point, the old Mapbox API still works, so I wouldn't break it!

**Comment by bampakoa** [comment]
@OmarShehata thanks very much for your feedback 😃 

> Can you merge in master and push?

No, I do not have permissions to do that in the CesiumJS repository

> Why is it called MapboxStyleProvider instead of MapboxStyleImageryProvider? It is an imagery provider, so it should be consistent with the other imagery providers listed here https://cesiumjs.org/Cesium/Build/Documentation/ImageryProvider.html?classFilter=ImageryProvid

You are right, I will update the PR

> It should probably be added to that list too I think ^

Sounds good to me. I will update the PR

> Since the old Mapbox API is deprecated, is there a reason to keep around the old MapboxImageryProvider? Can this new class just replace the old one?

I think that we shouldn't because as @mrlubos points out the old API has not been deprecated yet. See [this](https://github.com/AnalyticalGraphicsInc/cesium/issues/4015#issuecomment-472863603) comment for more info.

**Comment by OmarShehata** [comment]
> Do you mean to rebase to the master?

A rebase isn't necessary, you just need to pull the latest `master` from this repository into your branch here. See https://help.github.com/en/articles/merging-an-upstream-repository-into-your-fork

> I think that we shouldn't because as @mrlubos points out the old API has not been deprecated yet. See this comment for more info.

Thanks for the explanation - I'm still a little bit confused about the difference between these two routes that both get raster tiles from Mapbox:

* https://docs.mapbox.com/api/maps/#maps
* https://docs.mapbox.com/api/maps/#retrieve-a-sprite-image-or-json

Is the idea that using the style API (the second one) you have more fine grained control over customizing the look of the map? If so, do you have an example of something you can do with this that you couldn't do before?

I'm worried it's still a little big confusing having 2 Mapbox imagery providers that seem like they do the same thing, but just pass different parameters/use a different route? Would it make sense to have one MapboxImageryProvider but have a constructor option to use the style API?

**Comment by bampakoa** [comment]
> Is the idea that using the style API (the second one) you have more fine grained control over customizing the look of the map? If so, do you have an example of something you can do with this that you couldn't do before?

@OmarShehata Yes, that is right. You can create whatever style you want in Mapbox Studio (such as [this ](https://blog.mapbox.com/design-your-own-game-of-thrones-inspired-map-397460f09a25)one which is GoT based) and then render it inside the client. [Here](https://docs.mapbox.com/help/how-mapbox-works/map-design/#styles-vs-classic-styles) is a brief overview of the differences between them and [here](https://blog.mapbox.com/use-your-mapbox-studio-styles-everywhere-e29ce5954e47) is an article that explains why Mapbox is going to deprecate the old map styles 

> Would it make sense to have one MapboxImageryProvider but have a constructor option to use the style API?

I think that it would be complicated to maintain such a code as the two implementations have different required properties in order to create an object for each (`mapId` vs `styleId`). If we have two different constructors we are giving our users the opportunity to use whichever method they have created their maps and are comfortable with.

**Comment by OmarShehata** [comment]
I see what you mean now (that GoT inspired map is awesome!) Thanks for the explanation. I think that's a good reason to have two separate classes if they are different APIs as you describe. 

I noticed that they never actually use the word "deprecate" in that blog post. It seems like it's still a supported way of getting map tiles. I would say go ahead and name the new provider `MapboxStyleImageryProvider`. 

And then @tfili can you do a final review and merge after that?

**Comment by bampakoa** [comment]
@OmarShehata what about the base layer picker control? Should I leave it as is, to use the initial API of Mapbox?

**Comment by OmarShehata** [comment]
Since they're functionally equivalent, I don't think it really matters, but perhaps switching to the newer API is a good idea, so if we did want to add any custom styles to the base layer picker in the future it would be easier to do that.

File: Source/Scene/MapboxStyleProvider.js [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+var defaultCredit = new Credit('&copy; <a href="https://www.mapbox.com/about/maps/">Mapbox</a> &copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a> <strong><a href="https://www.mapbox.com/map-feedback/">Improve this map</a></strong>');
+
+/**
+ * Provides tiled imagery hosted by Mapbox.
+ *
+ * @alias MapboxStyleProvider
+ * @constructor
+ *
+ * @param {Object} [options] Object with the following properties:
+ * @param {String} [options.url='https://api.mapbox.com/styles/v1/'] The Mapbox server url.
```
Comment (tfili) [inline Source/Scene/MapboxStyleProvider.js]: This should be a `String` or a `Resource` like all our terrain/imagery providers.

Comment (bampakoa) [reply Source/Scene/MapboxStyleProvider.js]: @tfili I kept it as `String` in order to be the same as the current `MapboxImageryProvider`. Should I change it?

**Comment by bampakoa** [comment]
@OmarShehata @tfili this is ready

File: Source/Scene/MapboxStyleProvider.js (hunk 1 above)
Comment (tfili) [reply Source/Scene/MapboxStyleProvider.js]: Yeah, all our providers take `Resource` objects now. The MapBox Imagery provider was an oversight.

**Comment by bampakoa** [comment]
@tfili I updated the type of the parameter. I think the code already takes resource into account 😉

File: Source/Scene/MapboxStyleProvider.js (hunk 1 above)
Comment (bampakoa) [reply Source/Scene/MapboxStyleProvider.js]: @tfili I updated the type of the parameter. I think the code already takes resource into account 😉

**Comment by cesium-concierge** [comment]
Thanks again for your contribution @bampakoa! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

**Comment by OmarShehata** [comment]
@tfili this one would be nice to get in for release since it's already ready!

**Comment by tfili** [comment]
Thanks @bampakoa. Sorry it took so long. Everything seems to work great.

**Comment by bampakoa** [comment]
@tfili my change has been mistaken added to the 1.58 version release notes in CHANGES.md. Could you please correct it or do you want me to open a PR for this?

**Comment by OmarShehata** [comment]
Oh, good catch @bampakoa ! I got it here https://github.com/AnalyticalGraphicsInc/cesium/pull/7973

**Comment by mramato** [comment]
I still don't understand why we need a 2 Mapbox providers in master?  Is there a good reason to use one over the other?

**Comment by OmarShehata** [comment]
@mramato Mapbox has two different APIs for creating and serving imagery. See https://github.com/AnalyticalGraphicsInc/cesium/pull/7698#issuecomment-494514606

From what I understand, the newer one allows you to do customize more, but the old workflow isnt deprecated and removing it in Cesiumjs would be a breaking change.

**Comment by mrlubos** [comment]
@OmarShehata Yesss. Do you know if this new imagery will support vector tiles? (as opposed to the raster tiles served by the previous version)

**Comment by mramato** [comment]
@OmarShehata I understand that, so is there a reason there wasn't jut an in-place update or are the two incompatible?  If the latter, why wasn't the old one deprecated so that we can remove it 3 months from now?  Does the old version provide user-accessible functionality that the new one does not?

**Comment by mrlubos** [comment]
@mramato These are questions for Mapbox, not for Cesium. Mapbox released a new API with breaking changes.

**Comment by OmarShehata** [comment]
@mramato 

> why wasn't the old one deprecated so that we can remove it 3 months from now

I think the initial confusion here was that the Mapbox static tiles API was deprecated in favor of the styles API. It looks like it is supported and will be supported for the foreseeable future. This is just supporting an additional workflow with different options. For example the raster (old) API lets you specify the image format: https://docs.mapbox.com/api/maps/#raster-tiles. The styles (new) API https://docs.mapbox.com/api/maps/#static-tiles doesn't have this option. I think deprecating and removing it in the CesiumJS side would be removing a feature that's still actively used by the community. 

@mrlubos 

> Do you know if this new imagery will support vector tiles? (as opposed to the raster tiles served by the previous version)

@bampakoa can correct me if I'm wrong here, but it looks like the vector tiles are rasterized on the server, and that's what CesiumJS renders. We do have a feature request to support vector tiles directly in Cesium here: https://github.com/AnalyticalGraphicsInc/cesium/issues/2132

**Comment by mrlubos** [comment]
@OmarShehata I understand Mapbox supports vector tiles directly - https://docs.mapbox.com/api/maps/#retrieve-vector-tiles

**Comment by OmarShehata** [comment]
> I understand Mapbox supports vector tiles directly

Right, it's CesiumJS that doesn't support rendering these vector tiles.

**Comment by mrlubos** [comment]
Is this an issue for a separate pull request @OmarShehata @hpinkos? Anything on the roadmap?

**Comment by OmarShehata** [comment]
@mrlubos yes this would be a separate feature from this pull request. You can see the full discussion here: https://github.com/AnalyticalGraphicsInc/cesium/issues/2132

**Pull request merged**